zxlive/dodo.py

```diff
@@ -0,0 +1,261 @@
+    
+    VertexType = {0:'BOUNDARY',1:'Z',2:'X'}
+    EdgeType   = {1:'SIMPLE',2:'HADAMARD'}
+    
+    strPrime = "\nConsider a ZX-diagram defined by the following spiders:\n\nlabel, type, phase\n"
+    for v in g.vertices(): strPrime += str(v) + ', ' + str(VertexType[g.type(v)]) + ', ' + str(g.phase(v)) + '\n'
+    
+    strPrime += "\nwith the following edges:\n\nsource,target,type\n"
+    for e in g.edges(): strPrime += str(e[0]) + ', ' + str(e[1]) + ', ' + str(EdgeType[g.edge_type(e)]) + '\n'
+    strPrime += '\n'
+    
+    #Follows the format...
+    #
+    #"""
+    #Consider a ZX-diagram defined by the following spiders:
+    #
+    #label, type, phase
+    #0, Z, 0.25
+    #1, Z, 0.5
+    #2, X, 0.5
+    #
+    #with the following edges:
+    #
+    #source,target,type
+    #0, 1, SIMPLE
+    #1, 2, HADAMARD
+    #
+    #"""
+    
+    return strPrime
+
+def text_to_speech(text):
+    """Generates an mp3 file reading the given text (via OpenAI Whisper)."""
+    client = openai.OpenAI(
+        api_key = API_KEY
+    )
+
+    response = client.audio.speech.create(
+        model="tts-1",
+        voice="nova",
+        input=text,
+    )
+    
+    response.stream_to_file(GET_MODULE_PATH() + "/temp/Dodo_Dave_latest.mp3")
+    os.system(GET_MODULE_PATH() + "/temp/Dodo_Dave_latest.mp3") #TEMP/TODO - THIS SHOULD BE USING A PROPER IN-APP AUDIO PLAYER RATHER THAN OS
+
+def record_audio(duration=5, sample_rate=44100):
+    recording = sd.rec(int(duration * sample_rate), samplerate=sample_rate, channels=2, dtype='int16')
+    sd.wait()
+    return recording
+
+def save_as_mp3(audio_data, sample_rate=44100):
+    file_path = GET_MODULE_PATH() + "/temp/user_query_latest.mp3"
+    audio_segment = AudioSegment(
+        data=np.array(audio_data).tobytes(),
+        sample_width=2,
+        frame_rate=sample_rate,
+        channels=2
+    )
+    audio_segment.export(file_path, format='mp3')
+    return file_path
+
+def transcribe_audio(file_path):
+    client = openai.OpenAI(api_key=API_KEY)
+    with open(file_path, "rb") as audio_file:
+        transcription = client.audio.transcriptions.create(
+            model="whisper-1",
+            file=audio_file,
+            language='en'
+        )
+    #print(f'Transcription: {transcription.text}') #TEMP
+    return transcription.text
+
+def speech_to_text():
+    sample_rate = 44100  # Sample rate in Hz
+    duration = 5  # Duration of recording in seconds
+    audio_data = record_audio(duration, sample_rate)
+    file_path = save_as_mp3(audio_data, sample_rate)
+    txt = transcribe_audio(file_path)
+    return txt
+
+def encode_image(image_path):
+    with open(image_path, "rb") as image_file:
+        return base64.b64encode(image_file.read()).decode('utf-8')
+        
+def get_image_prompt():
+    """Returns the prompt that encourages DODO-GPT to describe the given image like a ZX-diagram."""
+
+    return """
```

Comment:
Maybe put this large string at the top of the file

Reply:
Sure!